RRF-87: Search Autocomplete

arianaantonio opened the pull request:
- Added temp API config
- Hooked up autocomplete and search to live data
- Added "brand" to RCHSearchProduct in SDK

arianaantonio committed:
RRF-87: Search Autocomplete
- Added temp API config
- Hooked up autocomplete and search to live data
- Added "brand" to RCHSearchProduct in SDK

arianaantonio committed:
RRF-87: Search Autocomplete
- Moved RCHAutocompleteSuggestion.h to public

KingOfBrian commented:
I'm getting `Error getting current user` when going to that screen. There's a chance that there's some UserDefaults configuration that you have set up that is having it work for you. Can you try a fresh install and see if it works? The Client configuration workaround isn't getting triggered because it's missing.

KingOfBrian commented on Demo/RCHSearchViewController.swift:
```diff
@@ -19,8 +19,8 @@ class RCHSearchViewController: UIViewController, UISearchBarDelegate, UICollecti
-    var productArray: [RCHRecommendedProduct] = []
-    var autocompleteArray: [String] = ["Boots", "Bomber", "Boucle", "Boyfriend"]
+    var productArray: [RCHSearchProduct] = []
+    var autocompleteArray: [String] = []
```
Don't include the type in the variable name. I'd use something like `products` and `autocompleteSuggestions` #pp

KingOfBrian commented on Demo/RCHSearchViewController.swift:
```diff
@@ -170,18 +205,18 @@ class RCHSearchViewController: UIViewController, UISearchBarDelegate, UICollecti
-                
+                let attributedString = NSMutableAttributedString(string: autocompleteString)
+            
+                let range: Range<String.Index> = autocompleteString.range(of: searchString)!
```
I got a crash on this a few times, not entirely sure what I was doing differently #req

KingOfBrian replied:
Typing fast seems to crash it. I think `searchString` is updated faster than the response.

arianaantonio committed:
RRF-87: Search Autocomplete
- Updated CurrentUser defaults

arianaantonio committed:
RRF-87: Search Autocomplete
- Cleaned up code
- Fixed bug with user defaults
- Fixed bug with fast typing in search bar

KingOfBrian commented:
You have a merge conflict @arianaantonio ! Also, The status bar should be updated to white inside the search app.

KingOfBrian commented:
Looks good @arianaantonio !
One thing we should probably do is put a 0.5 second delay on triggering both requests so we don't bog the system down. These can be triggered via timer in didSet on `searchTerm`.

 A few polish suggestions to take the pattern we talked about further: (#pp)
- Order the methods / vars -- State Management, API Calls, IBActions, UI Delegates.
   - I like to have my IBActions and UI Delegates just modify VC state or trigger API calls, and not have any conditionals or direct view modification. This isn't an absolute, a few small things is fine, but if things get bigger, it's best to model.
- Have the `searchTerm` property manage the text in the `UISearchBar`
- Have the attributed string returned in a method like `attributedText(for row: Int) -> NSAttributedString`

arianaantonio committed:
RRF-87: Search Autocomplete
- Cleaned up code
- Changed status bar color to white
- Added a timer for search API call

KingOfBrian commented on Demo/RCHSearchViewController.swift:
```diff
@@ -19,30 +19,56 @@ class RCHSearchViewController: UIViewController, UISearchBarDelegate, UICollecti
+    var searchTerm = "" {
+        didSet {
+            searchBar.text = searchTerm
+            var _ = Timer.scheduledTimer(timeInterval: 0.5, target: self, selector: #selector(self.searchForProducts), userInfo: nil, repeats: false)
```
We also want to save the timer and cancel the last one that was submitted so we don't just delay the search, but only do one search 0.5 seconds after the last keypress.

KingOfBrian commented on Demo/RCHSearchViewController.swift:
```diff
@@ -52,94 +78,109 @@ class RCHSearchViewController: UIViewController, UISearchBarDelegate, UICollecti
     
-    @IBAction func handleTapOnFooter(sender: UITapGestureRecognizer) {
-        resetSearch()
+    func resetSearch() {
```
`dismissSearch` may be a better name.

KingOfBrian commented on Demo/RCHSearchViewController.swift:
```diff
@@ -52,94 +78,109 @@ class RCHSearchViewController: UIViewController, UISearchBarDelegate, UICollecti
+            
+            self.autocompleteSuggestions = responseAutocompleteSuggestions.map({$0.text!})
+        }) { (responseObject, error) in
+            print(error)
```
We should present an alert if there's an error with the `localizedDescription` of the error as the message.

KingOfBrian commented on Demo/RCHSearchViewController.swift:
```diff
@@ -52,94 +78,109 @@ class RCHSearchViewController: UIViewController, UISearchBarDelegate, UICollecti
+        } else {
+            searchTerm = searchText
+            autocomplete(withQuery: searchText)
+        }
```
This entire behavior can get pushed into `searchTerm.didSet` I think.

arianaantonio replied:
I tried this before but if I move autocomplete into searchTerm.didSet then it gets called when the user selects an autocomplete suggestion as the searchTerm. Which means the autocomplete table pops up over the search results, and we don't want that here.

KingOfBrian commented on Demo/RCHSearchViewController.swift:
```diff
@@ -52,94 +78,109 @@ class RCHSearchViewController: UIViewController, UISearchBarDelegate, UICollecti
             return
         }
-        searchForProducts(withTerm: searchTerm)
+        searchTerm = searchText
```
Since we're searching as we go, I don't think we have to update the searchTerm here.

KingOfBrian reviewed (changes requested):
Few comments, I think the timer is the only #req item. Looks great!

arianaantonio committed:
RRF-87: Search autocomplete
- Handled canceling timer
- Consolidated some code

KingOfBrian commented on Demo/RCHSearchViewController.swift:
```diff
@@ -150,22 +163,12 @@ extension RCHSearchViewController: UISearchBarDelegate, UICollectionViewDelegate
-            autocomplete(withQuery: searchText)
-        }
+        searchTerm = searchText
+        autocomplete(withQuery: searchText)
```
I think we might want to do autocomplete on a timer too.
